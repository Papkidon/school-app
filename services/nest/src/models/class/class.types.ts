import { z } from 'zod';

export const classCreateSchema = z
  .object({
    data: z.object({
      number: z.number().int().min(0),
      max_population: z.number().int().min(0),
    }),
  })
  .strict();

export const classUpdateSchema = classCreateSchema.deepPartial();

export type classesWithSubjectsDto = {
  id: string;
  number: number;
  max_population: number;
  ClassSubject?: [
    {
      class_id: string;
      subject_id: string;
      subject?: {
        id: string;
        name: string;
        points: number;
      };
    },
  ];
}[];

export type mappedClassesWithSubjects = {
  id: string;
  number: number;
  max_population: number;
  subjects: {
    id: string;
    name: string;
    points: number;
  }[];
}[];

export type classType = {
  data: {
    number: number;
    max_population: number;
  };
};
