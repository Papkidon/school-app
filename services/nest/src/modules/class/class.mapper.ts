import {
  classesWithSubjectsDto,
  mappedClassesWithSubjects,
} from '../../models/class/class.types';

export const mapClassesWithSubjects = (
  classes: classesWithSubjectsDto,
): mappedClassesWithSubjects => {
  if (!classes.length) return [];
  return classes.map((subject) => {
    return {
      id: subject.id,
      number: subject.number,
      max_population: subject.max_population,
      subjects: subject.ClassSubject.map((clas) => {
        return {
          id: clas.subject.id,
          name: clas.subject.name,
          points: clas.subject.points,
        };
      }),
    };
  });
};
