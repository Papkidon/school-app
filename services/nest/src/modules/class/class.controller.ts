import {
  Body,
  Controller,
  Delete,
  Get,
  HttpException,
  HttpStatus,
  Param,
  Patch,
  Post,
} from '@nestjs/common';
import { Class } from '@prisma/client';
import {
  classCreateSchema,
  classType,
  classUpdateSchema,
} from '../../models/class/class.types';
import { DeepPartial, idSchema } from '../../models/utility.types';
import { ZodValidationPipe } from '../../pipes/zod.validation.pipe';
import { mapClassesWithSubjects } from './class.mapper';
import { ClassService } from './class.service';

@Controller('api/v1/class')
export class ClassController {
  constructor(private readonly classService: ClassService) {}

  @Get('')
  async findAll(): Promise<Class[]> {
    try {
      const result = await this.classService.findAll();

      if (!Object.keys(result).length) {
        throw new HttpException(
          'No class data exists in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Get('subjects')
  async findAllClassesWithSubjects() {
    try {
      const result = mapClassesWithSubjects(
        await this.classService.findAllClassesWithSubjects(),
      );

      if (!result.length) {
        throw new HttpException(
          'No class data exists in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Post('')
  async create(
    @Body(new ZodValidationPipe(classCreateSchema)) createClassDto: classType,
  ): Promise<Class> {
    try {
      const foundClass = await this.classService.findByNumber(
        createClassDto.data.number,
      );

      if (foundClass) {
        throw new HttpException(
          'Class with given number already exists',
          HttpStatus.CONFLICT,
        );
      }

      return await this.classService.create(createClassDto);
    } catch (error) {
      throw error;
    }
  }

  @Get(':id')
  async findById(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<Class> {
    try {
      const result = await this.classService.findById(id);

      if (!result) {
        throw new HttpException(
          'User with given id does not exist in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Patch(':id')
  async update(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
    @Body(new ZodValidationPipe(classUpdateSchema))
    body: DeepPartial<classType>,
  ): Promise<Class> {
    try {
      const foundClass = await this.classService.findById(id);

      if (!foundClass) {
        throw new HttpException(
          'Class with given id does not exist in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return await this.classService.update(id, body);
    } catch (error) {
      throw error;
    }
  }

  @Delete(':id')
  async delete(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<Class> {
    try {
      const foundClass = await this.classService.findById(id);

      if (!foundClass) {
        throw new HttpException(
          'Class with given id does not exist in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return await this.classService.delete(id);
    } catch (error) {
      throw error;
    }
  }
}
