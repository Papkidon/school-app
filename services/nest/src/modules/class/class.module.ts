import { Module } from '@nestjs/common';
import { ClassController } from './class.controller';
import { ClassService } from './class.service';
import { ClassRepository } from './class.repository';
import { LoggerModule } from 'nestjs-pino';
import { PrismaModule, PrismaService } from 'nestjs-prisma';

@Module({
  imports: [
    LoggerModule.forRoot({
      pinoHttp: {
        transport: {
          target: 'pino-pretty',
          options: {
            singleLine: true,
          },
        },
      },
    }),
    PrismaModule.forRoot(),
  ],
  controllers: [ClassController],
  providers: [ClassService, ClassRepository, PrismaService],
})
export class ClassModule {}
