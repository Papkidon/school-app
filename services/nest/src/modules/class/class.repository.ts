import { Injectable } from '@nestjs/common';
import { PrismaService as PrismaService } from 'nestjs-prisma';
import { classType } from '../../models/class/class.types';
import { DeepPartial } from '../../models/utility.types';
import { Class } from '@prisma/client';
import { classesWithSubjectsDto } from '../../models/class/class.types';

@Injectable()
export class ClassRepository {
  constructor(private prisma: PrismaService) {}
  async findAll(): Promise<Class[]> {
    return await this.prisma.class.findMany();
  }

  async findById(id: string): Promise<Class> {
    return await this.prisma.class.findUnique({ where: { id: id } });
  }

  async findAllClassesWithSubjects(): Promise<classesWithSubjectsDto> {
    return (await this.prisma.class.findMany({
      include: {
        ClassSubject: {
          include: {
            subject: true,
          },
        },
      },
    })) as unknown as classesWithSubjectsDto;
  }

  async create(body: any): Promise<Class> {
    return await this.prisma.class.create(body);
  }

  async delete(id: string): Promise<Class> {
    return await this.prisma.class.delete({ where: { id: id } });
  }

  async update(id: string, body: DeepPartial<classType>): Promise<Class> {
    return await this.prisma.class.update({
      where: { id: id },
      data: { ...body.data },
    });
  }
}
