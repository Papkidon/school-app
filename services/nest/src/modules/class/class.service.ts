import { Injectable } from '@nestjs/common';
import { ClassRepository } from './class.repository';
import { classType } from '../../models/class/class.types';
import { DeepPartial } from '../../models/utility.types';
import { Class } from '@prisma/client';
import { classesWithSubjectsDto } from '../../models/class/class.types';

@Injectable()
export class ClassService {
  constructor(private classRepository: ClassRepository) {}
  async findAll(): Promise<Class[]> {
    return await this.classRepository.findAll();
  }

  async findById(id: string): Promise<Class> {
    return await this.classRepository.findById(id);
  }

  async findByNumber(number: number): Promise<Class> {
    return await this.classRepository.findByNumber(number);
  }

  async findAllClassesWithSubjects(): Promise<classesWithSubjectsDto> {
    return await this.classRepository.findAllClassesWithSubjects();
  }

  async create(body: any): Promise<Class> {
    return await this.classRepository.create(body);
  }

  async delete(id: string): Promise<Class> {
    return await this.classRepository.delete(id);
  }

  async update(id: string, body: DeepPartial<classType>): Promise<Class> {
    return await this.classRepository.update(id, body);
  }
}
