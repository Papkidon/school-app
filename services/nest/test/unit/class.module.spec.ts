import { Class } from '@prisma/client';
import { PrismaService } from 'nestjs-prisma';
import {
  classesWithSubjectsDto,
  mappedClassesWithSubjects,
} from '../../src/models/class/class.types';
import { ClassController } from '../../src/modules/class/class.controller';
import * as classTypes from '../../src/modules/class/class.mapper';
import { ClassRepository } from '../../src/modules/class/class.repository';
import { ClassService } from '../../src/modules/class/class.service';
import {
  dummyClassCreate,
  dummyClassId,
  dummyClassUpdate,
} from './dummies/dummy.data';

describe('UserController', () => {
  let classController: ClassController;
  let classService: ClassService;
  let classRepository: ClassRepository;
  let prismaService: PrismaService;

  beforeEach(() => {
    prismaService = new PrismaService();
    classRepository = new ClassRepository(prismaService);
    classService = new ClassService(classRepository);
    classController = new ClassController(classService);
  });

  afterAll(() => {
    jest.useRealTimers();
    jest.resetAllMocks();
  });

  describe('findAll', () => {
    it('should return an array of all classes', async () => {
      const result = ['class1', 'class2'] as unknown as Class[];
      jest
        .spyOn(classRepository, 'findAll')
        .mockImplementation(async () => result);

      expect(await classController.findAll()).toMatchSnapshot();
    });
  });

  describe('findById', () => {
    it('should return a class with given id', async () => {
      const result = 'dummyFoundByIdClass' as unknown as Class;
      jest
        .spyOn(classRepository, 'findById')
        .mockImplementation(async () => result);

      expect(await classController.findById(dummyClassId)).toMatchSnapshot();
    });
  });

  describe('findClassesWithSubjects', () => {
    it('should return all classes with their subjects', async () => {
      const mappedResult =
        'dummyMappedResult' as unknown as mappedClassesWithSubjects;
      const result =
        'dummyFoundClassesWithSubjects' as unknown as classesWithSubjectsDto;

      jest
        .spyOn(classRepository, 'findAllClassesWithSubjects')
        .mockImplementation(async () => result);

      jest
        .spyOn(classTypes, 'mapClassesWithSubjects')
        .mockImplementation(() => mappedResult);

      expect(
        await classController.findAllClassesWithSubjects(),
      ).toMatchSnapshot();
    });
  });

  describe('create', () => {
    it('should create a new class', async () => {
      const dummyClassNotFoundByNumber = '' as unknown as Class;
      const result = 'dummyCreatedClass' as unknown as Class;
      jest
        .spyOn(classRepository, 'create')
        .mockImplementation(async () => result);

      jest
        .spyOn(classRepository, 'findByNumber')
        .mockImplementation(async () => dummyClassNotFoundByNumber);

      expect(await classController.create(dummyClassCreate)).toMatchSnapshot();
    });
  });

  describe('update', () => {
    it('should update a class with given id', async () => {
      const dummyFoundClassById = 'dummyFoundClassById' as unknown as Class;
      const result = 'dummyUpdatedClass' as unknown as Class;
      jest
        .spyOn(classRepository, 'update')
        .mockImplementation(async () => result);

      jest
        .spyOn(classRepository, 'findById')
        .mockImplementation(async () => dummyFoundClassById);

      expect(
        await classController.update(dummyClassId, dummyClassUpdate),
      ).toMatchSnapshot();
    });
  });

  describe('delete', () => {
    it('should delete a class with given id', async () => {
      const dummyFoundClassById = 'dummyFoundClassById' as unknown as Class;
      const result = 'dummyDeletedUser' as unknown as Class;
      jest
        .spyOn(classRepository, 'delete')
        .mockImplementation(async () => result);

      jest
        .spyOn(classRepository, 'findById')
        .mockImplementation(async () => dummyFoundClassById);

      expect(await classController.delete(dummyClassId)).toMatchSnapshot();
    });
  });
});
