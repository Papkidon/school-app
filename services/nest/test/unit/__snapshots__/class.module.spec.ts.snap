// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserController create should create a new class 1`] = `"dummyCreatedClass"`;

exports[`UserController delete should delete a class with given id 1`] = `"dummyDeletedUser"`;

exports[`UserController findAll should return an array of all classes 1`] = `
[
  "class1",
  "class2",
]
`;

exports[`UserController findById should return a class with given id 1`] = `"dummyFoundByIdClass"`;

exports[`UserController findClassesWithSubjects should return all classes with their subjects 1`] = `"dummyMappedResult"`;

exports[`UserController update should update a class with given id 1`] = `"dummyUpdatedClass"`;
